### PR TITLE
feat: shared workspaces

### PR DESCRIPTION
**Pull request by Gankra**
This introduces a new axoproject WorkspaceGraph abstraction that exists to replace the old Workspaces approach. The purpose of this new abstraction is that it supports nested workspaces, while still allowing you to work with all the packages uniformly.

Most code largely remains the same, as it looks up properties of packages by id, or queries a property of all workspaces/packages, where essentially nothing is changed.

Also all logic to inherit config from workspaces to packages is just changed to inherit from the *root* package, as we don't want to support multiple `[workspace.metadata.dist]` entries.

compute_build functions have been changed to take a workspace, so that we try to compute N cargo builds for N cargo workspaces (discarding them if unecessary).

The URL code has once again been refactored to be more of a chain instead of redoing the same work over and over.

The `package_info` subfield of a workspace has been deprecated in favour of Asking the WorkspaceGraph, as there are different tiers of childhood now.

### CODE REVIEW

**Comment by Gankra** [comment]
This builds and passes tests locally, but there are several outstanding TODOs, and this isn't *really* tested until we change the gallery.

File: axoproject/src/lib.rs [hunk 1]
```diff
@@ -33,6 +33,186 @@ mod tests;
...
+            .map(|(idx, pkg)| (PackageIdx(idx), pkg))
+    }
+
+    /// Get all workspaces
+    pub fn all_workspace_indices(&self) -> impl Iterator<Item = WorkspaceIdx> {
+        (0..self.workspaces.len()).map(WorkspaceIdx)
+    }
+
+    /// Try to get a consensus repository_url for the given packages (or all if None given)
+    pub fn repository_url(&self, packages: Option<&[PackageIdx]>) -> Result<Option<RepositoryUrl>> {
```
Comment (mistydemeo) [inline axoproject/src/lib.rs:204]: Wondering if it also makes sense to keep a version of this on individual workspaces. Guess we're probably not looking at that individually as often.

**Review by mistydemeo — COMMENTED** [review]
Generally, I like this. The `WorkspaceGraph` interface is nice and clean, and I appreciate how it looks / gets used within cargo-dist. I'm mainly wondering about the couple bits of pseudo-private data and what the plan is for those.

File: axoproject/src/lib.rs (hunk 1 above)
Comment (Gankra) [reply axoproject/src/lib.rs:204]: Yeah you can always get that by just querying all packages under the workspace.

File: axoproject/src/lib.rs [hunk 2]
```diff
@@ -119,18 +313,21 @@ pub enum WorkspaceKind {
...
 pub struct WorkspaceInfo {
-    /// The kinf of workspace this is (Rust or Javascript)
+    /// The kind of workspace this is (Rust or Javascript)
     pub kind: WorkspaceKind,
     /// The directory where build output will go (generally `target/`)
     pub target_dir: Utf8PathBuf,
     /// The root directory of the workspace (where the root Cargo.toml is)
     pub workspace_dir: Utf8PathBuf,
-    /// Computed info about the packages.
+    /// Nested workspaces (exists temp, do not use outside axoproject)
```
[the remark this answers is not in the document]
Comment (Gankra) [reply axoproject/src/lib.rs:324]: Yeah this is basically a hack I did to get things working without melting my brain...

File: axoproject/src/lib.rs [hunk 3]
```diff
@@ -229,6 +413,14 @@ impl WorkspaceInfo {
 /// specify their location -- something Cargo does but Guppy (and cargo-metadata) don't.
 #[derive(Debug)]
 pub struct PackageInfo {
+    /// The unoverrideable name of the package within its own package management system.
+    ///
+    /// This may be needed when e.g. talking to cargo about the package.
+    pub true_name: String,
```
Comment (mistydemeo) [inline axoproject/src/lib.rs:419]: Smart plan.

**Review by mistydemeo — APPROVED** [review]

**Pull request merged**